crates/lair_keystore_api/src/sodium_secretstream.rs

```diff
@@ -192,72 +241,123 @@ where
+                let mut eph_cbox_pub =
+                    [0; sodoken::crypto_box::XSALSA_PUBLICKEYBYTES];
+                let mut eph_cbox_sec = sodoken::SizedLockedArray::<
+                    { sodoken::crypto_box::XSALSA_SECRETKEYBYTES },
+                >::new()?;
+                sodoken::crypto_box::xsalsa_keypair(
+                    &mut eph_cbox_pub,
+                    &mut eph_cbox_sec.lock(),
+                )?;
+
+                Ok((eph_cbox_pub, eph_cbox_sec))
+            })
+            .await
+            .map_err(OneErr::new)??;
 
         // generate an ephemeral kx keypair
-        let eph_kx_pub = sodoken::BufWriteSized::new_no_lock();
-        let eph_kx_sec = sodoken::BufWriteSized::new_mem_locked()?;
-        kx::keypair(eph_kx_pub.clone(), eph_kx_sec.clone())?;
+        let (eph_kx_pub, mut eph_kx_sec) =
+            tokio::task::spawn_blocking(move || -> LairResult<_> {
+                let mut eph_kx_pub =
+                    [0; sodoken::crypto_box::XSALSA_PUBLICKEYBYTES];
+                let mut eph_kx_sec = sodoken::SizedLockedArray::<
+                    { sodoken::crypto_box::XSALSA_SECRETKEYBYTES },
+                >::new()?;
+                sodoken::crypto_box::xsalsa_keypair(
+                    &mut eph_kx_pub,
+                    &mut eph_kx_sec.lock(),
+                )?;
+
+                Ok((eph_kx_pub, eph_kx_sec))
+            })
+            .await
+            .map_err(OneErr::new)??;
 
         // sealed initiator message
         let mut message: [u8; 64] = [0; 64];
-        message[..32].copy_from_slice(&eph_cbox_pub.read_lock());
-        message[32..].copy_from_slice(&eph_kx_pub.read_lock());
-        let message = sodoken::BufReadSized::from(message);
-        let cipher =
-            <sodoken::BufWriteSized<{ 64 + cbox::SEALBYTES }>>::new_no_lock();
-        cbox::seal(cipher.clone(), message, srv_id_pub_key).await?;
-        let cipher = cipher.try_unwrap_sized().unwrap();
+        message[..32].copy_from_slice(&eph_cbox_pub);
+        message[32..].copy_from_slice(&eph_kx_pub);
+
+        let mut cipher =
+            tokio::task::spawn_blocking(move || -> LairResult<_> {
+                let mut cipher = sodoken::SizedLockedArray::<
+                    { 64 + sodoken::crypto_box::XSALSA_SEALBYTES },
+                >::new()?;
+                sodoken::crypto_box::xsalsa_seal(
+                    &mut *cipher.lock(),
+                    &message,
+                    &srv_id_pub_key,
+                )?;
+
+                Ok(cipher)
+            })
+            .await
+            .map_err(OneErr::new)??;
 
         // write the sealed initiator
-        send.write_all(&cipher).await?;
+        send.write_all(&*cipher.lock()).await?;
 
         // read the sealed response ephemeral kx pub key
-        let mut cipher: [u8; 32 + cbox::SEALBYTES] = [0; 32 + cbox::SEALBYTES];
+        // Todo Blank cipher on unseal??
```

Comment:
This looks wrong to me. The cipher argument to `seal_open` and `xsalsa_seal_open` are both read-only, so they aren't filling this value in.

Reply:
the [recv.read_exact](https://github.com/holochain/lair/blob/e82937521ae9b7bdb30c8b0736c13cd4220a0223/crates/lair_keystore_api/src/sodium_secretstream.rs#L234-L235) below is filling in the cipher to be read by seal_open

Reply:
Wow... that is RIGHT THERE! I've been staring at this code too many hours today clearly